services/src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -41,61 +44,60 @@
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicLong;
+import scala.collection.JavaConversions;
 
 
-/**
- * This sets up the producers used by Kafka Monitoring and some of the reported metrics.
- */
 public class ProduceService implements Service {
   private static final Logger LOG = LoggerFactory.getLogger(ProduceService.class);
   private static final String METRIC_GROUP_NAME = "produce-service";
 
   private final String _name;
   private final ProduceMetrics _sensors;
-  private final KMBaseProducer _producer;
-  private final ScheduledExecutorService _executor;
+  private volatile KMBaseProducer _producer;
+  private volatile ScheduledExecutorService _executor;
+  private final ScheduledExecutorService _checkAddPartitionsExecutor;
```

Comment:
I tried using the existing thread pool and having the thread coordinate with each other.  There is a way to make that correct but it is complicated by:
1) only wanting one of N threads to check periodically
2) coordinating with all the running threads
3) checking if we need a new CyclicBarrier

1) We basically need to recreate what the schedule thread executor provides already, but we want one of the scheduled threads to do this at a potentially different interval;  you probably don't want to poll zookeeper every 100ms. This needs to set some flag for threads to wait on the barrier (see 2).
2) All the threads need to stop so that the producer can be swapped out with a new producer.  This means introducing a CyclicBarrier.  When all the expected threads hit the barrier the Runnable associated with the barrier will trigger which will change the producer.
3) The CyclicBarrier does not actually trigger on a fixed number of threads.  The number of active threads is not actually equal to the number of threads in the executor service.  It's actually min(number of partitions, number of threads).  So the cyclic barrier needs to be switched out.

Just using a separate thread in a different executor service that is scheduled to check for added partitions saves all this complexity.